File: apps/uploader/src/uploader/domain/models/sequencing.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Analysis:
    analysis_identifier: str | None = None
    belongs_to_sequencing: str | None = None
    physical_data_location: str | None = None
    abstract_data_location: str | None = None
    data_formats_stored: list[str] | None = None
    algorithms_used: list[str] | None = None
    reference_genome_used: str | None = None
    bioinformatic_protocol_used: str | None = None
    bioinformatic_protocol_deviation: str | None = None
    reason_for_bioinformatic_protocol_deviation: str | None = None
    wgs_guideline_followed: str | None = None


@dataclass(frozen=True)
class Sequencing:
    sequencing_identifier: str | None = None
    belongs_to_sample_preparation: str | None = None
    sequencing_date: str | None = None
    sequencing_platform: str | None = None
    sequencing_instrument_model: str | None = None
    sequencing_method: str | None = None
    median_read_depth: int | None = None
    observed_read_length: int | None = None
    observed_insert_size: int | None = None
    percentage_q30: float | None = None
    percentage_tr20: float | None = None
    other_quality_metrics: str | None = None
    analysis: Analysis | None = None


@dataclass(frozen=True)
class SamplePreparation:
    sampleprep_identifier: str | None = None
    belongs_to_material: str | None = None
    input_amount: str | None = None
    library_preparation_kit: str | None = None
    pcr_free: bool | None = None
    target_enrichment_kit: str | None = None
    full_sequence_genes: list[str] | None = None
    partial_sequence_genes: list[str] | None = None
    umis_present: bool | None = None
    intended_insert_size: int | None = None
    intended_read_length: int | None = None
    sequencing: Sequencing | None = None


@dataclass(frozen=True)
class SequencingEntry:
    predictive_number: str
    source_id: str
    fixed_block_identifier: str | None = None
    sample_preparation: SamplePreparation | None = None


SequencingData = list[SequencingEntry]

File: src/Uploader/Uploader.Domain/Sequencing.cs
using Uploader.Domain.Common;

namespace Uploader.Domain;

/// <summary>FAIR Genomes <c>Analysis</c> value object.</summary>
public sealed record Analysis : ValueObject
{
    public string? AnalysisIdentifier { get; init; }
    public string? BelongsToSequencing { get; init; }
    public string? PhysicalDataLocation { get; init; }
    public string? AbstractDataLocation { get; init; }
    public IReadOnlyList<string>? DataFormatsStored { get; init; }
    public IReadOnlyList<string>? AlgorithmsUsed { get; init; }
    public string? ReferenceGenomeUsed { get; init; }
    public string? BioinformaticProtocolUsed { get; init; }
    public string? BioinformaticProtocolDeviation { get; init; }
    public string? ReasonForBioinformaticProtocolDeviation { get; init; }
    public string? WgsGuidelineFollowed { get; init; }
}

/// <summary>FAIR Genomes <c>Sequencing</c> value object.</summary>
public sealed record Sequencing : ValueObject
{
    public string? SequencingIdentifier { get; init; }
    public string? BelongsToSamplePreparation { get; init; }
    public string? SequencingDate { get; init; }
    public string? SequencingPlatform { get; init; }
    public string? SequencingInstrumentModel { get; init; }
    public string? SequencingMethod { get; init; }
    public int? MedianReadDepth { get; init; }
    public int? ObservedReadLength { get; init; }
    public int? ObservedInsertSize { get; init; }
    public double? PercentageQ30 { get; init; }
    public double? PercentageTr20 { get; init; }
    public string? OtherQualityMetrics { get; init; }
    public Analysis? Analysis { get; init; }
}

/// <summary>FAIR Genomes <c>SamplePreparation</c> value object.</summary>
public sealed record SamplePreparation : ValueObject
{
    public string? SampleprepIdentifier { get; init; }
    public string? BelongsToMaterial { get; init; }
    public string? InputAmount { get; init; }
    public string? LibraryPreparationKit { get; init; }
    public bool? PcrFree { get; init; }
    public string? TargetEnrichmentKit { get; init; }
    public IReadOnlyList<string>? FullSequenceGenes { get; init; }
    public IReadOnlyList<string>? PartialSequenceGenes { get; init; }
    public bool? UmisPresent { get; init; }
    public int? IntendedInsertSize { get; init; }
    public int? IntendedReadLength { get; init; }
    public Sequencing? Sequencing { get; init; }
}

/// <summary>One sequencing entry keyed by predictive number.</summary>
public sealed record SequencingEntry : Entity
{
    public required string PredictiveNumber { get; init; }
    public required string SourceId { get; init; }
    public string? FixedBlockIdentifier { get; init; }
    public SamplePreparation? SamplePreparation { get; init; }
}
